### ISSUE TITLE
Negative CTC loss

### ISSUE THREAD

**Issue by Jzmo** [espnet/espnet#863]
Hey guys,

I am working on AMI multichannel recipe, and encounter a negative CTC loss issue. In training process the CTC loss sometimes appear to be nan, large positive and negative value as shown below:
`2019-06-13 02:30:01,545 (e2e_asr:302) WARNING: loss (=nan) is not correct`
`2019-06-13 03:52:24,416 (e2e_asr:302) WARNING: loss (=14484257.000000) is not correct`
`2019-06-13 04:22:57,220 (e2e_asr:302) WARNING: loss (=93660798500470784.000000) is not correct`
`2019-06-13 05:06:23,975 (e2e_asr:302) WARNING: loss (=-186341) is not correct`
`2019-06-13 05:26:13,287 (e2e_asr:302) WARNING: loss (=-61910442439.000000) is not correct`

Some observations:
1. Both warp-ctc and build in ctc report this issue
2. Issue dose not disappear as iteration goes
3. Utterances which cause this warning are not same in every epoch
4. When training without ctc loss it works well
5. After delete utterance less than 0.1s (prevent bad fbank features) the issue is still there
6. When change the number of mel-frequency bins (decrease to 40) the issue is still there 

Any one have any ideas about this?

**Comment by ShigekiKarita** [comment]
When the text is shorter than its subsampled speech, ctc loss results nan. Maybe 0.1s is not enough. Anyway, this warning does no harm because our implementation simply ignores this awful loss value before `loss.backward()`.

**Comment by sw005320** [comment]
I have never had the negative CTC loss in the single channel system. Always positive. This seems to be a specific issue in the multichannel system. Besides that, I'm concerning about the current utterance wise BF algorithm because some of the AMI utterances are too short to compute the BF filter.

**Comment by Jzmo** [comment]
Thanks for your reply and advice. I'm now train the network by just ignoring of negative loss. However, there is a new observation confused me a lot: When running on Tesla K80 GPU, the number of bad loss value is much less than running on GTX 1080ti.
For example for the first 600 iterations on GTX 1080ti:
```
epoch       iteration   main/loss   main/loss_ctc  main/loss_att 
0           100         183.56      221.355        145.766
2019-06-16 22:12:15,718 (e2e_asr:302) WARNING: loss (=nan) is not correct
2019-06-16 22:12:40,496 (e2e_asr:302) WARNING: loss (=nan) is not correct
0           200         145.946     178.686        113.205
2019-06-16 22:14:55,883 (e2e_asr:302) WARNING: loss (=-196890.328125) is not correct
2019-06-16 22:15:21,816 (e2e_asr:302) WARNING: loss (=nan) is not correct
2019-06-16 22:15:26,280 (e2e_asr:302) WARNING: loss (=nan) is not correct
0           300         138.009     175.188        100.829
2019-06-16 22:15:41,076 (e2e_asr:302) WARNING: loss (=nan) is not correct
2019-06-16 22:15:43,954 (e2e_asr:302) WARNING: loss (=nan) is not correct
2019-06-16 22:16:00,873 (e2e_asr:302) WARNING: loss (=nan) is not correct
2019-06-16 22:16:04,120 (e2e_asr:302) WARNING: loss (=nan) is not correct
0           400         109.493     142.072        76.9144
0           500         126.898     164.395        89.4002
```
... and on Tesla K80:
```
epoch       iteration   main/loss   main/loss_ctc  main/loss_att
0           100         178.4       213.722        143.079
0           200         143.909     176.454        111.364
0           300         134.269     170.549        97.989
0           400         104.998     135.863        74.1333
0           500         126.808     164.207        89.4103
```
Do you have any idea about this phenomenon?

**Comment by b-flo** [comment]
Hi,

I observed the same phenomenon when training with CTC a single channel system on a French corpus. In my case it only occured when using VGG bottleneck and was mainly due to the fact the corpus contains irregular segment-utterance pairs.
However as @ShigekiKarita said it, it didn't harm the training.

**Comment by sw005320** [comment]
> When running on Tesla K80 GPU, the number of bad loss value is much less than running on GTX 1080ti.

This is very interesting.
Could you test the same setup at the single channel (normal `asr1`) recipe?
I just want to make sure this happens from CTC or something in the multichannel recipe.

Best,
Shinji

**Comment by Jzmo** [comment]
> Could you test the same setup at the single channel (normal asr1) recipe?
I did this test and the result seems indicate is not only appear in multichannel recipe.
The normal asr1 recipe has less awful loss. But it reports 4 times bad loss when training on K80 with in 10000 iterations while 14 times on 1080ti.

**Comment by sw005320** [comment]
Thanks. I just want to make sure that you observed the negative CTC loss even for normal `asr`? Then, this is probably due to some transcription issues in AMI.

**Comment by Jzmo** [comment]
Yes there are some negative CTC loss for normal `asr` but much less than in multichannle reicpe, For example
```
2019-06-18 11:35:26,545 (e2e_asr:302) WARNING: loss (=-2.190914) is not correct
2019-06-18 11:36:14,774 (e2e_asr:302) WARNING: loss (=-1.187396) is not correct
2019-06-18 12:25:59,884 (e2e_asr:302) WARNING: loss (=-1018850262216430682546416211984384.000000) is not correct
```

**Comment by sw005320** [comment]
OK. Then, let's try to continue the training for both single- and multi-channel systems by skipping the negative CTC loss.

**Comment by sw005320** [comment]
@jnishi or @HawkAaron, do you know why such large negative values happen?  I used to observe large positive values, which I believe comes from the initial values for loss, but I have no idea about the negative large numbers...

**Comment by sw005320** [comment]
Hi @Jzmo, my friend told me that softmax used in CTC may produce `0` and this would be replaced by the large negative number in https://github.com/espnet/espnet/blob/83101a26177500542d1fcff9c22bf7752c2f49ad/espnet/nets/pytorch_backend/e2e_asr.py#L137.
Could you check to see whether this happen?
